integration/bridge.go

```diff
@@ -179,3 +182,59 @@ func (bc *bridgeConn) close() {
 	bc.in.Close()
 	bc.out.Close()
 }
+
+func (b *bridge) Blackhole() {
+	b.mu.Lock()
+	close(b.blackholec)
+	b.mu.Unlock()
+}
+
+func (b *bridge) Unblackhole() {
+	b.mu.Lock()
+	b.blackholec = make(chan struct{})
+	b.mu.Unlock()
+}
+
+// https://github.com/golang/go/blob/master/src/io/io.go copyBuffer
+func ioCopy(dst io.Writer, src io.Reader, blackholec chan struct{}) (written int64, err error) {
+	// If the reader has a WriteTo method, use it to do the copy.
+	// Avoids an allocation and a copy.
+	if wt, ok := src.(io.WriterTo); ok {
+		return wt.WriteTo(dst)
+	}
+	// Similarly, if the writer has a ReadFrom method, use it to do the copy.
+	if rt, ok := dst.(io.ReaderFrom); ok {
+		return rt.ReadFrom(src)
+	}
```

Comment:
can drop the above ReaderFrom/WriterTo stuff since it's optimizations ignoring the blackhole channel